File: base64_url.rb
# base64_url.rb — Ruby версия

require 'base64'
require 'clipboard'
require 'optparse'

options = {}
OptionParser.new do |opts|
  opts.banner = "Usage: ruby base64_url.rb [options]"
  opts.on("--input STRING", "Входная строка") { |v| options[:input] = v }
  opts.on("--file FILE", "Файл для обработки") { |v| options[:file] = v }
  opts.on("--decode", "Декодировать") { |v| options[:decode] = true }
  opts.on("--output FILE", "Сохранить результат в файл") { |v| options[:output] = v }
  opts.on("--no-copy", "Не копировать в буфер обмена") { |v| options[:no_copy] = true }
end.parse!

puts "\e[36m🔐 Base64 URL-safe Encoder (Ruby)\e[0m"

result = ""

if options[:file]
  puts "📂 Обработка файла: #{options[:file]}"
  content = File.read(options[:file], encoding: 'BINARY')
  if options[:decode]
    result = Base64.urlsafe_decode64(content).force_encoding('UTF-8')
  else
    result = Base64.urlsafe_encode64(content, padding: false)
  end
elsif options[:input]
  puts "📝 Входные данные: #{options[:input]}"
  if options[:decode]
    result = Base64.urlsafe_decode64(options[:input])
  else
    result = Base64.urlsafe_encode64(options[:input], padding: false)
  end
else
  puts "📝 Чтение из STDIN (Ctrl+D для окончания)"
  data = STDIN.read
  if data.empty?
    puts "\e[33m⚠️ Пустой ввод.\e[0m"
    exit 1
  end
  if options[:decode]
    result = Base64.urlsafe_decode64(data)
  else
    result = Base64.urlsafe_encode64(data, padding: false)
  end
end

puts "\e[32mРезультат:\e[0m"
puts result
puts

unless options[:no_copy]
  begin
    Clipboard.copy(result)
    puts "\e[32m✅ Результат скопирован в буфер обмена!\e[0m"
  rescue
    puts "\e[33m⚠️ Не удалось скопировать в буфер обмена.\e[0m"
  end
end

if options[:output]
  File.write(options[:output], result)
  puts "\e[32m💾 Сохранено в #{options[:output]}\e[0m"
end

File: base64_url.py


### 1. `base64_url.py` (Python)

```python
# base64_url.py — Python версия

import sys
import base64
import argparse
import os
import subprocess
import platform
from pathlib import Path

try:
    import pyperclip
    HAS_CLIPBOARD = True
except ImportError:
    HAS_CLIPBOARD = False

def copy_to_clipboard(text):
    if HAS_CLIPBOARD:
        pyperclip.copy(text)
        return True
    system = platform.system()
    if system == "Windows":
        subprocess.run(["clip"], input=text.encode("utf-8"), check=True, capture_output=True)
        return True
    elif system == "Darwin":
        subprocess.run(["pbcopy"], input=text.encode("utf-8"), check=True)
        return True
    elif system == "Linux":
        try:
            subprocess.run(["xclip", "-selection", "clipboard"], input=text.encode("utf-8"), check=True)
            return True
        except FileNotFoundError:
            try:
                subprocess.run(["xsel", "-b"], input=text.encode("utf-8"), check=True)
                return True
            except FileNotFoundError:
                pass
    return False

def encode_string(data):
    return base64.urlsafe_b64encode(data.encode("utf-8")).decode("ascii").rstrip("=")

def encode_file(filename):
    with open(filename, "rb") as f:
        return base64.urlsafe_b64encode(f.read()).decode("ascii").rstrip("=")

def decode_string(data):
    # Добавляем padding, если нужно
    padding = 4 - (len(data) % 4)
    if padding != 4:
        data += "=" * padding
    return base64.urlsafe_b64decode(data).decode("utf-8")

def decode_file(filename):
    with open(filename, "r") as f:
        content = f.read().strip()
    return decode_string(content)

def main():
    parser = argparse.ArgumentParser(description="Base64 URL-safe Encoder")
    parser.add_argument("input", nargs="?", help="Входная строка или путь к файлу")
    parser.add_argument("--file", "-f", help="Файл для кодирования/декодирования")
    parser.add_argument("--decode", "-d", action="store_true", help="Декодировать")
    parser.add_argument("--output", "-o", help="Сохранить результат в файл")
    parser.add_argument("--no-copy", "-n", action="store_true", help="Не копировать в буфер обмена")
    args = parser.parse_args()

    print("\033[36m🔐 Base64 URL-safe Encoder (Python)\033[0m")

    result = None
    if args.file:
        print(f"📂 Обработка файла: {args.file}")
        if args.decode:
            result = decode_file(args.file)
        else:
            result = encode_file(args.file)
    elif args.input:
        print(f"📝 Входные данные: {args.input}")
        if args.decode:
            result = decode_string(args.input)
        else:
            result = encode_string(args.input)
    else:
        print("📝 Чтение из STDIN (Ctrl+D для окончания)")
        content = sys.stdin.read()
        if not content:
            print("⚠️ Пустой ввод.")
            sys.exit(1)
        if args.decode:
            result = decode_string(content)
        else:
            result = encode_string(content)

    if result is None:
        sys.exit(1)

    print(f"\033[32mРезультат:\033[0m")
    print(result)
    print()

    if not args.no_copy:
        if copy_to_clipboard(result):
            print("\033[32m✅ Результат скопирован в буфер обмена!\033[0m")
        else:
            print("\033[33m⚠️ Не удалось скопировать в буфер обмена.\033[0m")

    if args.output:
        with open(args.output, "w", encoding="utf-8") as f:
            f.write(result)
        print(f"\033[32m💾 Сохранено в {args.output}\033[0m")

if __name__ == "__main__":
    main()
